src/pydvl/utils/parallel/futures/ray.py

```diff
@@ -0,0 +1,343 @@
+                # to the work queue
+                if self.work_queue.full():
+                    return
+                try:
+                    work_item = self.pending_queue.get_nowait()
+                except queue.Empty:
+                    logger.debug("pending queue is empty")
+                    return
+                else:
+                    if work_item is None:
+                        self.work_queue.put_nowait(work_item)
+                        return
+                    logger.debug("moving work item from pending queue to work queue")
+                    self.work_queue.put(work_item)
+                    del work_item
+
+    def submit_work_item(self) -> None:
+        with self.queue_lock:
+            logger.debug("work item manager thread acquired queue lock")
+            # Try to get a work item
+            # If it is None, we break from the loop.
+            # Otherwise, we submit the future.
+            try:
+                logger.debug("getting work item from work queue")
+                work_item = self.work_queue.get_nowait()
+                if work_item is None:
+                    return
+            except queue.Empty:
+                logger.debug("work queue is empty")
+                return
+        logger.debug("Submitting work item")
+        work_item.run()
+        self.submitted_futures.add(work_item.future)
+        # Delete references to object
+        del work_item
+
+    def is_shutting_down(self) -> bool:
+        # Check whether we should start shutting down the executor.
+        executor = self.executor_reference()
+        # No more work items can be added if:
+        #   - The executor that owns this worker has been collected OR
+        #   - The executor that owns this worker has been shutdown.
+        with self.shutdown_lock:
+            logger.debug("work item manager thread acquired shutdown lock")
+            if executor is None or executor._shutdown:
+                # Flag the executor as shutting down as early as possible if it
+                # is not gc-ed yet.
+                if executor is not None:
+                    executor._shutdown = True
+                return True
+            return False
+
+    def flag_executor_shutting_down(self):
+        # Flag the executor as shutting down and cancel remaining tasks if
+        # requested as early as possible if it is not gc-ed yet.
+        executor = self.executor_reference()
+        with self.shutdown_lock:
+            logger.debug("work item manager thread acquired shutdown lock")
+            if executor is not None:
+                executor._shutdown = True
+                # Cancel pending work items if requested.
+                if executor._cancel_pending_futures:
+                    logger.debug("forcefully cancelling running futures")
+                    # We cancel the future's object references
+                    # We cannot cancel a running future object.
```

Comment:
`ray.cancel(force=True)` cancels the task on the ray cluster but `Future` objects cannot be cancelled if their status is set to `RUNNING`. 

When calling `ray.cancel()` the future will terminate and its result will be an exception.

This is the code of the `cancel()` method:

```python
    def cancel(self):
        """Cancel the future if possible.

        Returns True if the future was cancelled, False otherwise. A future
        cannot be cancelled if it is running or has already completed.
        """
        with self._condition:
            if self._state in [RUNNING, FINISHED]:
                return False

            if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:
                return True

            self._state = CANCELLED
            self._condition.notify_all()

        self._invoke_callbacks()
        return True
```